src/v/ssx/include/ssx/single_sharded.h

```diff
@@ -152,11 +153,14 @@ class single_sharded : ss::sharded<maybe_service<Service>> {
         return base::invoke_on(
           _shard,
           options,
-          [func = std::forward<Func>(func)](
-            maybe_service<Service>& maybe_service, Args&&... args) {
-              func(*maybe_service, std::forward<Args>(args)...);
-          },
-          std::forward<Args>(args)...);
+          [func_tup = std::forward_as_tuple(std::forward<Func>(func)),
+           args_tup = std::tuple(std::move(args)...)](
+            maybe_service<Service>& maybe_service) mutable {
+              return std::apply(
+                std::forward<Func>(std::get<0>(func_tup)),
+                std::tuple_cat(
+                  std::forward_as_tuple(*maybe_service), std::move(args_tup)));
+          });
```

Comment:
this could use a comment explaining things for those of us that are mere c++ mortals 😄 .

at first glance, though, `func_tup = std::forward_as_tuple(std::forward<Func>(func))` seems problematic for cases when I want to transfer ownership of a function instance since, IIUC, this is capturing references that will used asynchronously to caller.

Reply:
Thanks @dotnwat, you are right, I have changed it, added comments, and changed the tests to check what happens if references provided to the call are invalidated before the future is awaited. The test did catch problems with both args and functions.